src/miniscript/analyzable.rs

```diff
@@ -22,7 +22,115 @@ use core::fmt;
 use std::error;
 
 use crate::prelude::*;
-use crate::{Miniscript, MiniscriptKey, ScriptContext};
+use crate::{Miniscript, MiniscriptKey, ScriptContext, Terminal};
+
+/// Params for parsing miniscripts that either non-sane or non-specified(experimental) in the spec.
+/// Used as a parameter [`Miniscript::from_str_insane`] and [`Miniscript::parse_insane`].
+///
+/// This allows parsing miniscripts if
+/// 1. It is unsafe(does not require a digital signature to spend it)
+/// 2. It contains a unspendable path because of either
+///     a. Resource limitations
+///     b. Timelock Mixing
+/// 3. The script is malleable and thereby some of satisfaction weight
+///    guarantees are not satisfied.
+/// 4. It has repeated public keys
+/// 5. raw pkh fragments without the pk. This could be obtained when parsing miniscript from script
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Default)]
+pub struct ExtParams {
+    /// Allow parsing of non-safe miniscripts
+    pub top_unsafe: bool,
+    /// Allow parsing of miniscripts with unspendable paths
+    pub resource_limitations: bool,
+    /// Allow parsing of miniscripts with timelock mixing
+    pub timelock_mixing: bool,
+    /// Allow parsing of malleable miniscripts
+    pub malleability: bool,
+    /// Allow parsing of miniscripts with repeated public keys
+    pub repeated_pk: bool,
+    /// Allow parsing of miniscripts with raw pkh fragments without the pk.
+    /// This could be obtained when parsing miniscript from script
+    pub raw_pkh: bool,
+}
+
+impl ExtParams {
+    /// Create a new ExtParams that with all the sanity rules
+    pub fn new() -> ExtParams {
+        ExtParams {
+            top_unsafe: false,
+            resource_limitations: false,
+            timelock_mixing: false,
+            malleability: false,
+            repeated_pk: false,
+            raw_pkh: false,
+        }
+    }
+
+    /// Create a new ExtParams that allows all the sanity rules
+    pub fn sane() -> ExtParams {
+        ExtParams::new()
+    }
+
+    /// Create a new ExtParams that insanity rules
+    pub fn insane() -> ExtParams {
+        ExtParams {
+            top_unsafe: true,
+            resource_limitations: true,
+            timelock_mixing: true,
+            malleability: true,
+            repeated_pk: true,
+            raw_pkh: false,
+        }
+    }
+
+    /// Enable all non-sane rules and experimental rules
+    pub fn all() -> ExtParams {
```

Comment:
In 89169c23aec0bac4009593e6c451bf9b0432ee6d:

Can you rename this `allow_all` or (somewhat snarkily) `free_for_all`?